shm/Ship.cpp

```diff
@@ -0,0 +1,88 @@
+#include "Ship.h"
+
+#include <algorithm>
+
+Ship::Ship(size_t capacity, size_t maxCrew, size_t speed, const std::string& name, size_t id, Time* time)
+    : capacity_(capacity), max_crew_(maxCrew), crew_(0), speed_(speed), name_(name), id_(id), time_(time) {
+    time_->subscribe(this);
+}
+
+//override from Observer;
+Ship::~Ship() {
+    time_->unsubscribe(this);
+}
+
+void Ship::setName(const std::string& name) {
+    name_ = name;
+}
+
+Ship& Ship::operator-=(size_t num) {
+    if (num >= crew_) {
+        crew_ = 0;
+    } else {
+        crew_ -= num;
+    }
+    return *this;
+}
+
+Ship& Ship::operator+=(size_t num) {
+    crew_ += num;
+    if (crew_ > max_crew_) {
+        crew_ = max_crew_;
+    }
```

Comment:
I think it's mistake to add before checking. If you add value that exceeds size_t maximum value you may get garbage value that's lower than crew limit.